server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/utils/StarTreeQueryHelper.java

```diff
@@ -0,0 +1,140 @@
+            && context.sort() == null
+            && (!trackTotalHits || context.trackTotalHitsUpTo() == SearchContext.TRACK_TOTAL_HITS_DISABLED)
+            && context.trackScores() == false
+            && context.minimumScore() == null
+            && context.terminateAfter() == 0;
+        return canUseStarTree;
+    }
+
+    /**
+     * Gets a parsed OriginalOrStarTreeQuery from the search context and source builder.
+     * Returns null if the query cannot be supported.
+     */
+    public static OriginalOrStarTreeQuery getOriginalOrStarTreeQuery(SearchContext context, SearchSourceBuilder source) throws IOException {
+        // Current implementation assumes only single star-tree is supported
+        CompositeDataCubeFieldType compositeMappedFieldType = (StarTreeMapper.StarTreeFieldType) context.mapperService()
+            .getCompositeFieldTypes()
+            .iterator()
+            .next();
+        CompositeIndexFieldInfo starTree = new CompositeIndexFieldInfo(
+            compositeMappedFieldType.name(),
+            compositeMappedFieldType.getCompositeIndexType()
+        );
+
+        StarTreeQuery starTreeQuery = StarTreeQueryHelper.toStarTreeQuery(starTree, compositeMappedFieldType, source.query());
+        if (starTreeQuery == null) {
+            return null;
+        }
+
+        for (AggregatorFactory aggregatorFactory : context.aggregations().factories().getFactories()) {
+            if (validateStarTreeMetricSuport(compositeMappedFieldType, aggregatorFactory) == false) {
+                return null;
+            }
+        }
+
+        return new OriginalOrStarTreeQuery(starTreeQuery, context.query());
+    }
+
+    private static StarTreeQuery toStarTreeQuery(
+        CompositeIndexFieldInfo starTree,
+        CompositeDataCubeFieldType compositeIndexFieldInfo,
+        QueryBuilder queryBuilder
+    ) {
+        Map<String, Long> queryMap;
+        if (queryBuilder == null || queryBuilder instanceof MatchAllQueryBuilder) {
+            queryMap = null;
+        } else if (queryBuilder instanceof TermQueryBuilder) {
+            List<String> supportedDimensions = compositeIndexFieldInfo.getDimensions()
+                .stream()
+                .map(Dimension::getField)
+                .collect(Collectors.toList());
+            queryMap = getStarTreePredicates(queryBuilder, supportedDimensions);
+        } else {
+            return null;
+        }
+
+        return new StarTreeQuery(starTree, queryMap);
+    }
+
+    /**
+     * Parse query body to star-tree predicates
+     * @param queryBuilder
+     * @return predicates to match
+     */
+    private static Map<String, Long> getStarTreePredicates(QueryBuilder queryBuilder, List<String> supportedDimensions) {
+        TermQueryBuilder tq = (TermQueryBuilder) queryBuilder;
+        String field = tq.fieldName();
+        if (!supportedDimensions.contains(field)) {
+            throw new IllegalArgumentException("unsupported field in star-tree");
```

Comment:
Lets not throw exception here , lets return null and handle it gracefully by not using star tree in this case

Reply:
Ack, we aren't throwing any exceptions any way. We suppress all exceptions related to star-tree at the moment and revert back to default flow. But can handle this scenario without throwing the exception as well.

```
try {
                OriginalOrStarTreeQuery parsedQuery = StarTreeQueryHelper.getOriginalOrStarTreeQuery(context, source);
                if (parsedQuery != null) {
                    context.parsedQuery(new ParsedQuery(parsedQuery));
                    logger.debug("can use star tree");
                } else {
                    logger.debug("cannot use star tree");
                }
            } catch (IOException ignored) {}
```

Reply:
As long as we are falling back to original query, we are good.

This failed by the way with the exception when we ran terms query against dimensions in the latest branch today. Maybe we can add one e2e test maybe on an index where field is not part of star tree dimensions.

Reply:
Sure, I'll make the change to not throw an exception anyway and add a test case for this validation.